src/geometry/euclidean/sphere.lean

```diff
@@ -0,0 +1,136 @@
+-/
+
+lemma mul_norm_eq_sub_norm_pow_two {x y z : V}
+  (h₁ : ∃ k : ℝ, k ≠ 1 ∧ x + y = k • (x - y)) (h₂ : ∥z - y∥ = ∥z + y∥) :
+  ∥x - y∥ * ∥x + y∥ = abs (∥z + y∥ ^ 2 - ∥z - x∥ ^ 2) :=
+begin
+  obtain ⟨k, hk_ne_one, hk⟩ := h₁,
+  let r := (k - 1)⁻¹ * (k + 1),
+
+  have hxy : x = r • y,
+  { rw [← smul_smul, eq_inv_smul_iff' (sub_ne_zero.mpr hk_ne_one), ← sub_eq_zero],
+    calc  (k - 1) • x - (k + 1) • y
+        = (k • x - x) - (k • y + y) : by simp_rw [sub_smul, add_smul, one_smul]
+    ... = (k • x - k • y) - (x + y) : by simp_rw [← sub_sub, sub_right_comm]
+    ... = k • (x - y) - (x + y)     : by rw ← smul_sub k x y
+    ... = 0                         : sub_eq_zero.mpr hk.symm },
+
+  have hzy : ⟪z, y⟫ = 0,
+  { rw [← eq_of_pow_two_eq_pow_two (norm_nonneg (z - y)) (norm_nonneg (z + y)),
+        norm_add_pow_two_real, norm_sub_pow_two_real] at h₂,
+    linarith },
+
+  have hzx : ⟪z, x⟫ = 0 := by rw [hxy, inner_smul_right, hzy, mul_zero],
+
+  symmetry,
+
+  calc  abs (∥z + y∥ ^ 2 - ∥z - x∥ ^ 2)
+      = abs (∥x∥ ^ 2 - ∥y∥ ^ 2)            : by simp [norm_add_pow_two_real, norm_sub_pow_two_real,
+                                                    hzy, hzx, abs_sub]
+  ... = abs (r ^ 2 * ∥y∥ ^ 2 - ∥y∥ ^ 2)    : by rw [hxy, norm_smul, mul_pow, norm_eq_abs, sqr_abs]
+  ... = abs ((r - 1) * (r + 1) * ∥y∥ ^ 2) : by ring_nf
+  ... = ∥r - 1∥ * ∥r + 1∥ * ∥y∥ ^ 2         : by simp [abs_mul, norm_eq_abs]
+  ... = ∥r - 1∥ * ∥y∥ * (∥r + 1∥ * ∥y∥)      : by ring
+  ... = ∥(r - 1) • y∥ * ∥(r + 1) • y∥      : by simp_rw [norm_smul]
+  ... = ∥x - y∥ * ∥x + y∥                  : by simp [sub_smul, add_smul, hxy],
+end
+
+end inner_product_geometry
+
+namespace euclidean_geometry
+
+/-!
+### Geometrical results on spheres in Euclidean affine spaces
+
+This section develops some results on spheres in Euclidean affine spaces.
+-/
+
+open inner_product_geometry
+
+variables {P : Type*} [metric_space P] [normed_add_torsor V P]
+include V
+
+/-- If `P` is a point on the line `AB` and `Q` is equidistant from `A` and `B`, then
+`AP * BP = abs (BQ ^ 2 - PQ ^ 2)`. -/
+lemma mul_dist_eq_sub_dist_pow_two {a b p q : P}
+  (hp : ∃ k : ℝ, k ≠ 1 ∧ b -ᵥ p = k • (a -ᵥ p)) (hq : dist a q = dist b q) :
+  dist a p * dist b p = abs (dist b q ^ 2 - dist p q ^ 2) :=
+begin
+  let m : P := midpoint ℝ a b,
+  obtain ⟨v, h1, h2, h3⟩ := ⟨vsub_sub_vsub_cancel_left, v a p m, v p q m, v a q m⟩,
+  have h : ∀ r, b -ᵥ r = (m -ᵥ r) + (m -ᵥ a) :=
+    λ r, by rw [midpoint_vsub_left, ← right_vsub_midpoint, add_comm, vsub_add_vsub_cancel],
+  iterate 4 { rw dist_eq_norm_vsub V },
+  rw [← h1, ← h2, h, h],
+  rw [← h1, h] at hp,
+  rw [dist_eq_norm_vsub V a q, dist_eq_norm_vsub V b q, ← h3, h] at hq,
+  exact mul_norm_eq_sub_norm_pow_two hp hq,
+end
+
+/-- If `A`, `B`, `C`, `D` are cospherical and `P` is on both lines `AB` and `CD`, then
+`AP * BP = CP * DP`. -/
+lemma mul_dist_eq_mul_dist_of_cospherical {a b c d p : P} (h : cospherical ({a, b, c, d} : set P))
```

Comment:
```suggestion
lemma mul_dist_eq_mul_dist {a b c d p : P} (h : cospherical ({a, b, c, d} : set P))
```

Reply:
I think the idea behind the naming is that the more descriptive the name, the better. I actually left the `cospherical`, instead of deleting it, and added `_and_smul` because of the smul condition: `mul_dist_eq_mul_dist_of_cospherical_and_smul`.